src/main/java/org/candlepin/controller/util/ContentPathBuilder.java

```diff
@@ -0,0 +1,188 @@
+        }
+
+        String fullPath = buildFullPath(environmentId, contentPath);
+        String populatedPath = populatePlaceholders(fullPath, environmentId);
+        return encodePath(populatedPath);
+    }
+
+    public String buildFullPath(String environmentId, String contentPath) {
+        String normalizedContentPath = normalizePath(contentPath);
+        if (isFullUri(normalizedContentPath)) {
+            return normalizedContentPath;
+        }
+
+        String envPrefix = this.envPrefixes.getOrDefault(environmentId, "");
+        String contentPathWithEnv = envPrefix + normalizedContentPath;
+        if (isFullUri(contentPathWithEnv)) {
+            return contentPathWithEnv;
+        }
+
+        return this.ownerPrefix + contentPathWithEnv;
+    }
+
+    private void add(Environment environment) {
+        String envPrefix = normalizePath(environment.getContentPrefix());
+        this.envPrefixes.put(environment.getId(), envPrefix);
+        // This should be a separate object, but we can extract it later,
+        // when we know if there are other placeholders we want to support.
+        this.envNames.put(environment.getId(), environment.getName());
+    }
+
+    private boolean isFullUri(String path) {
+        Matcher matcher = PROTOCOL_PATTERN.matcher(path);
+        return matcher.matches();
+    }
+
+    private String normalizePath(String path) {
+        if (path == null || path.isBlank()) {
+            return "";
+        }
+
+        String strippedPath = StringUtils.stripStart(StringUtils.stripEnd(path, "/"), "/");
+        if (isFullUri(strippedPath)) {
+            return strippedPath;
+        }
+
+        return "/" + strippedPath;
+    }
+
+    private String populatePlaceholders(String path, String envId) {
+        return path.replace("$env", this.envNames.getOrDefault(envId, ""));
+    }
+
+    private String encodePath(String contentPath) {
+        if (!isFullUri(contentPath)) {
+            return "/" + encodePathSegments(contentPath);
+        }
+        URIBuilder uriBuilder = urlBuilder(contentPath);
+        String scheme = uriBuilder.getScheme();
+        String host = uriBuilder.getHost();
+        String port = uriBuilder.getPort() == -1 ? "" : ":" + uriBuilder.getPort();
+        String path = uriBuilder.getPath();
+        String encodedPath = encodePathSegments(path);
+        return "%s://%s%s/%s".formatted(scheme, host, port, encodedPath);
+    }
+
+    private String encodePathSegments(String path) {
+        if (path == null || path.isBlank()) {
+            return "";
+        }
+        return Arrays.stream(path.split("/"))
+            .filter(s -> !s.isBlank())
+            .map(this::encode)
+            .collect(Collectors.joining("/"));
+    }
+
+    private String encode(String pathSegment) {
+        if (pathSegment.contains("$releasever")) {
+            return encodeSkippkingPlaceholder(pathSegment, "$releasever");
+        }
+        if (pathSegment.startsWith("$basearch")) {
```

Comment:
Why is one of these a `.contains` and one a `.startsWith`?

Also, given how it operates, I wonder if we couldn't condense a number of these methods down to one or two methods with a recursive, regex-based parser. Something to the effect of:

```
private String encode(String pathSegment) {
  Pattern parserPattern = Pattern.compile("^([^$]*)(\\$[A-Za-z]+)?(.*)$"); // const, probably
  Matcher matcher = parserPattern.matcher(pathSegment);

  if (matcher.matches()) {
    StringBuilder builder = new StringBuilder()
      .append(Util.encodeUrl(matcher.group(1)))
      .append(this.encodeMacro(matcher.group(2)));

    String remaining = matcher.group(3);
    if (!remaining.isEmpty()) {
      builder.append(this.encode(remaining));
    }

    return builder.toString();
  }
  else {
    return Util.encodeUrl(pathSegment);
  }
}

private String encodeMacro(String macro) {
  // if macro is null, return an empty string
  // if macro is a known macro ($env), convert it to the appropriate value, encoding as necessary
  // otherwise, if macro is unknown, assume it is passthrough and return it as-is ($basearch, $releasever)
}
```

Reply:
Another thing to consider here: One of the things we'll be doing with the entirety of the templating work will be using placeholders in the path that will be passed all the way back to the CDN server. For instance: `https://example.cdn.com/templates/$template/<actual content path here>`.

URL encoding would convert that to `%24template`, but the CDN server would probably convert that back? What are your thoughts on this? Depending on how we want to handle it, the macro encoding step probably needs to change accordingly (specifically that "unknown macro" bit).